File: rust/src/bin/p0079.rs
// Project Euler: Problem 79

/*
  I saw the Wikipedia page about topological sorting.
    https://en.wikipedia.org/wiki/Topological_sorting

  Note: This implementation finds only one topological sort not all.
*/

use std::collections::HashMap;

euler::run_solver!(79);

static FILE_DATA: &str = include_str!("../../assets/0079_keylog.txt");

fn solve() -> String {
    compute(FILE_DATA).to_string()
}

fn compute(data: &str) -> u64 {
    use euler::math;

    let mut graph = parse_data(data);
    let mut acc: Vec<u64> = Vec::new();
    let keys: Vec<u64> = graph.keys().copied().collect();

    for v in keys {
        acc = dfs(&mut graph, &mut acc, v);
    }
    acc.reverse();

    math::undigits(&acc)
}

fn parse_data(data: &str) -> HashMap<u64, Vec<u64>> {
    let mut ret: HashMap<u64, Vec<u64>> = HashMap::new();

    for line in data.lines() {
        let v: Vec<_> = line
            .chars()
            .map(|ch| u64::from(ch.to_digit(10).unwrap()))
            .collect();
        ret.entry(v[0]).or_default().extend(vec![v[1], v[2]]);
        ret.entry(v[1]).or_default().push(v[2]);
    }

    ret
}

fn dfs(graph: &mut HashMap<u64, Vec<u64>>, perm: &mut [u64], v: u64) -> Vec<u64> {
    fn visit(temp: &[u64], visited: &[u64], node: u64, graph: &HashMap<u64, Vec<u64>>) -> Vec<u64> {
        assert!(!temp.contains(&node), "a cycle path is found");

        if visited.contains(&node) {
            return visited.to_vec();
        }

        if let Some(lst) = graph.get(&node) {
            let mut acc = visited.to_vec();
            for &v in lst {
                acc = visit(&[vec![node], temp.to_vec()].concat(), &acc, v, graph);
            }
            [vec![node], acc].concat()
        } else {
            vec![node]
        }
    }

    visit(&Vec::new(), perm, v, graph)
}

#[cfg(test)]
mod tests {
    use super::compute;

    #[test]
    fn p0079() {
        assert_eq!(compute(super::FILE_DATA), 73162890);
    }
}

File: go/internal/sol/p0079/p0079.go
package p0079

import (
	_ "embed"
	"maps"
	"slices"
	"strings"
)

//go:embed 0079_keylog.txt
var fileContent string

// topological sorting
func dfs(graph map[string][]string, perm []string, v string) []string {
	var visit func([]string, []string, string) []string
	visit = func(temp []string, visited []string, node string) []string {
		if slices.Contains(temp, node) {
			panic("cycle detection")
		}
		if slices.Contains(visited, node) {
			return visited
		}
		if _, ok := graph[node]; ok {
			acc := visited
			for v := range slices.Values(graph[node]) {
				acc = visit(slices.Concat([]string{node}, temp), acc, v)
			}
			return slices.Concat([]string{node}, acc)
		} else {
			return []string{node}
		}
	}

	return visit([]string{}, perm, v)
}

func parseData(data string) map[string][]string {
	m := make(map[string][]string)

	for line := range strings.SplitSeq(strings.Trim(data, "\n"), "\n") {
		x := strings.Split(line, "")
		for kv := range slices.Values([][]string{{x[0], x[1]}, {x[0], x[2]}, {x[1], x[2]}}) {
			if v, ok := m[kv[0]]; ok {
				m[kv[0]] = append(v, kv[1])
			} else {
				m[kv[0]] = []string{kv[1]}
			}
		}
	}

	// sort & dedup
	for k, xs := range maps.All(m) {
		slices.Sort(xs)
		i := 0
		for j := 1; j < len(xs); j++ {
			if xs[i] == xs[j] {
				continue
			}
			i++
			xs[i] = xs[j]
		}
		m[k] = xs[:i+1]
	}

	return m
}

func compute(data string) string {
	graph := parseData(data)

	var acc []string
	for vertex := range maps.Keys(graph) {
		acc = dfs(graph, acc, vertex)
	}

	var result string
	for s := range slices.Values(acc) {
		result += s
	}

	return result
}

func Solve() string {
	return compute(fileContent)
}
